pkg/controller/service_instance.go

```diff
@@ -0,0 +1,183 @@
+// parametersFrom and forcing updateRequest to higher value when the referenced secrets change. The annotation
+// stores the hex value of the bcryted hash of the sha256 hash of the contents of all secrets referenced in
+// parametersFrom.
+//
+// The annotation is only ever added to the spec object not the actual object. The modified spec is treated as
+// if the user manually set the annotation.  This means that if the spec is updated to remove all parametersFrom
+// referenced secrets the annotation will be removed.
+//
+// If the a user actually did set the annotation manually there are two cases.
+// 1. if there are no parametersFrom referenced secrets we leave the spec untouched
+// 2. if there are referenced secrets we proceed to check and generate a new value for the annotation
+//
+// This can be disabled by adding the annotation with the value 'disabled'
+func (st *resourceSyncTask) processServiceInstance(spec *unstructured.Unstructured, actual runtime.Object, namespace string) (specRet *unstructured.Unstructured, e error) {
+	var instanceSpec sc_v1b1.ServiceInstance
+	var previousEncodedChecksum string
+	var updateCount int64
+
+	if err := unstructured_conversion.DefaultConverter.FromUnstructured(spec.Object, &instanceSpec); err != nil {
+		return nil, errors.Wrapf(err, "Failure to parse ServiceInstance spec %q", spec.GetName())
+	}
+
+	if len(instanceSpec.Spec.ParametersFrom) == 0 {
+		return spec, nil
+	}
+
+	if v, ok := instanceSpec.ObjectMeta.Annotations[secretParametersChecksumAnnotation]; ok && v == disabled {
+		return spec, nil
+	}
+
+	if actual != nil {
+		spec, ok := actual.(*sc_v1b1.ServiceInstance)
+		if !ok {
+			return nil, errors.New("Failure retrieving ServiceInstance spec")
+		}
+		previousEncodedChecksum = spec.ObjectMeta.Annotations[secretParametersChecksumAnnotation]
+		updateCount = spec.Spec.UpdateRequests
+	}
+
+	checkSum, err := st.calculateNewCheckSum(&instanceSpec, namespace, previousEncodedChecksum, updateCount)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to generate new checksum for spec %q", spec.GetName())
+	}
+
+	if previousEncodedChecksum != "" && checkSum != previousEncodedChecksum {
+		instanceSpec.Spec.UpdateRequests = updateCount + 1
+	}
+
+	instanceSpec = setAnnotation(instanceSpec, checkSum)
+
+	unstructuredSpec, err := util.RuntimeToUnstructured(&instanceSpec)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to marshal back into unstructed %q", spec.GetName())
+	}
+
+	return unstructuredSpec, nil
+}
+
+func (st *resourceSyncTask) calculateNewCheckSum(instanceSpec *sc_v1b1.ServiceInstance, namespace string, previousEncodedChecksum string, updateCount int64) (string, error) {
+	checksumPayload, err := st.generateSecretParametersChecksumPayload(&instanceSpec.Spec, namespace)
+	if err != nil {
+		return "", err
+	}
+
+	if previousEncodedChecksum != "" {
+		previousCheckSum, err := decodeChecksum(previousEncodedChecksum)
```

Comment:
Invalid data in the annotation should not stop the processing - just ignore the error and overwrite the invalid value.

Reply:
Maybe log the error